### PR TITLE
SFTP bug fixes

### PR DESCRIPTION
**Pull request by bagajjal**
1) bug #479 - "ls c:" is not working
   sanitized_path() is modified to handle the edge case "\x:"

2) bug #476 - "cd c:" is not working
   If "c:" is passed to _fullpath() then it is returning existing path but not "c:", so if we append "\\" to "c:" then it is working fine.

3) bug #474 - code cleanup MAX_PATH
   Using PATH_MAX variable instead of MAX_PATH
   In dirent.h, used PATH_MAX instead of hardcoding 256 characters
   In readdir(), changed the pdirentry to be a static variable. Before this, we are leaking the memory.

4) bug #467 - SFTP rename failed if the newpath already exists.
    The _wrename throws error if the newpath exists.
	To make it consistent with the linux behavrior,
     a) if the newpath is a file and if it exists then delete it so that _wrename will succeed.
	 b) if the newpath is a directory and if it is empty then delete it so that _wrename will succeed.

### CODE REVIEW

File: contrib/win32/win32compat/misc.c [hunk 1]
```diff
@@ -532,6 +532,26 @@ w32_rename(const char *old_name, const char *new_name) {
 		return -1;
 	}
 
+	// To be consistent with linux rename(),
+	// 1) if the new_name is file, then delete it so that _wrename will succeed.
```
Comment (manojampalam) [inline contrib/win32/win32compat/misc.c]: use /* comments */

File: contrib/win32/win32compat/misc.c [hunk 2]
```diff
@@ -532,6 +532,26 @@ w32_rename(const char *old_name, const char *new_name) {
...
+	// 2) if the new_name is directory and it is empty then delete it so that _wrename will succeed.
+	struct _stat64 st;
+	if (fileio_stat(sanitized_path(new_name), &st) != -1) {
+		if(((st.st_mode & _S_IFMT) == _S_IFREG)) {
+			w32_unlink(new_name);
+		} else {
+			DIR *dirp = opendir(new_name);
+			if (NULL != dirp) {
+				struct dirent *dp = readdir(dirp);
```
Comment (manojampalam) [inline contrib/win32/win32compat/misc.c:548]: free(dp) ?

File: contrib/win32/win32compat/misc.c [hunk 3]
```diff
@@ -658,15 +678,20 @@ convertToForwardslash(char *str) {
...
-realpath(const char *path, char resolved[MAX_PATH]) {
-	char tempPath[MAX_PATH];
+realpath(const char *path, char resolved[PATH_MAX]) {
+	char tempPath[PATH_MAX];
 		
-	if ( (strlen(path) >= 2) && (path[0] == '/') && (path[2] == ':') )
+	if ((path[0] == '/') && path[1] && (path[2] == ':')) {
 		strncpy(resolved, path + 1, strlen(path)); // skip the first '/'
-	else
+		if (strlen(resolved) == 2) { // make "x:" as "x:/"
```
Comment (manojampalam) [inline contrib/win32/win32compat/misc.c]: Take it out of parent if. to take care of x: input (expected output x:/)

File: contrib/win32/win32compat/misc.c [hunk 4]
```diff
@@ -676,6 +701,19 @@ realpath(const char *path, char resolved[MAX_PATH]) {
 	return resolved;
 }
 
+char*
+sanitized_path(const char *path) {
+	char tempPath[PATH_MAX];
+	if (realpath(path, tempPath)) {
```
Comment (manojampalam) [inline contrib/win32/win32compat/misc.c]: get rid of realpath to avoid unnecessary return value handling

**Review by manojampalam — CHANGES_REQUESTED** [review]

File: contrib/win32/win32compat/misc.c (hunk 2 above)
Comment (bagajjal) [reply contrib/win32/win32compat/misc.c:548]: as per our discussion.. readdir() will return the static pointer so free(dp) is not required here..

**Review by manojampalam — APPROVED** [review]

**Pull request merged**